### PR TITLE
CONSOLE-2456: Update OperatorHub detail view to include default catalog sources

### PR DESCRIPTION
**Pull request by harishsurf**
This change adds the default catalog sources section to the operator hub "Detail" view (Cluster Settings -> Global Configuration -> OperatorHub)  providing an option to enable/disable each of the default catalog sources through a modal as shown in the image below

toggle

[CONSOLE-2456](https://issues.redhat.com/browse/CONSOLE-2456)

Signed-off-by: Harish <hgovinda@redhat.com>

### CODE REVIEW

File: frontend/public/components/utils/details-item.tsx [hunk 1]
```diff
@@ -98,7 +99,9 @@ export const DetailsItem: React.FC<DetailsItemProps> = ({
             <>
               <SplitItem isFilled />
               <SplitItem>
-                <EditButton onClick={onEdit}>{t('public~Edit')}</EditButton>
+                <EditButton data-test={`details-item-value__${label}`} onClick={onEdit}>
```
Comment (TheRealJon) [inline frontend/public/components/utils/details-item.tsx:102]: [nit] This test id doesn't seem to really match the EditButton component. Should be something like `${label}-details-item__edit-button`. Also I would recommend just passing label through to `EditButton` as a prop, then defining the data-test attribute on the `Button` within that component. You'll also need to update the type definition for `EditButton` props.  
```suggestion
                <EditButton id={label} onClick={onEdit}>
```

File: frontend/public/components/utils/details-item.tsx [hunk 2]
```diff
@@ -39,7 +39,7 @@ export const PropertyPath: React.FC<{ kind: string; path: string | string[] }> =
 const EditButton: React.SFC<{ onClick: (e: React.MouseEvent<HTMLButtonElement>) => void }> = (
```
Comment (TheRealJon) [inline frontend/public/components/utils/details-item.tsx:39]:
```suggestion
const EditButton: React.FC<{ onClick: (e: React.MouseEvent<HTMLButtonElement>) => void, id: string }> = (
```

File: frontend/public/components/utils/details-item.tsx [hunk 3]
```diff
@@ -39,7 +39,7 @@ export const PropertyPath: React.FC<{ kind: string; path: string | string[] }> =
...
   props,
 ) => (
-  <Button variant="link" isInline onClick={props.onClick}>
+  <Button variant="link" isInline onClick={props.onClick} {...props}>
```
Comment (TheRealJon) [inline frontend/public/components/utils/details-item.tsx:42]:
```suggestion
  <Button variant="link" isInline onClick={props.onClick} data-test={`${id}-details-item__edit-button`}>
```

File: frontend/packages/operator-lifecycle-manager/src/components/operator-hub/operator-hub-details.tsx [hunk 4]
```diff
@@ -6,19 +6,52 @@ import {
...
+            showNodeSelector={false}
+          />
+        </div>
+        <div className="col-sm-6 col-xs-12">
+          <div className="co-m-pane__details">
+            <DetailsItem
+              label="Default Sources"
+              obj={operatorHub}
+              path="status.sources"
+              canEdit
```
Comment (TheRealJon) [inline frontend/packages/operator-lifecycle-manager/src/components/operator-hub/operator-hub-details.tsx:37]: @spadgett Do we need to do an access check here to make sure the user can edit the OperatorHub resource?

**Review by TheRealJon — APPROVED** [review]
Just a few small tweaks

File: frontend/public/components/utils/details-item.tsx (hunk 1 above)
Comment (harishsurf) [reply frontend/public/components/utils/details-item.tsx:102]: added a `label` prop

File: frontend/public/components/utils/details-item.tsx (hunk 2 above)
Comment (harishsurf) [reply frontend/public/components/utils/details-item.tsx:39]: extracted out the props to a `type EditButtonProps = { }`

File: frontend/packages/operator-lifecycle-manager/src/components/operator-hub/operator-hub-details.tsx (hunk 4 above)
Comment (spadgett) [reply frontend/packages/operator-lifecycle-manager/src/components/operator-hub/operator-hub-details.tsx:37]: Good catch! Yes, let's add an access review.

Comment (harishsurf) [reply frontend/packages/operator-lifecycle-manager/src/components/operator-hub/operator-hub-details.tsx:37]: Added the check

File: frontend/packages/operator-lifecycle-manager/src/components/modals/edit-default-sources-modal.tsx [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+  return (
+    <form onSubmit={submit} name="form" className="modal-content modal-content--no-inner-scroll">
+      <ModalTitle>Edit Default Sources</ModalTitle>
+      <ModalBody>
+        {operatorHub.status.sources.map((source) => {
+          return (
+            <React.Fragment key={source.name}>
+              <div className="row co-m-form-row">
+                <div className="col-xs-5">
```
[the remark this answers is not in the document]
Comment (harishsurf) [reply frontend/packages/operator-lifecycle-manager/src/components/modals/edit-default-sources-modal.tsx]: I had issues in aligning the label and the toggle icon in the edit-default-sources-modal.tsx. I have raised a [patternfly issue](https://github.com/patternfly/patternfly-react/issues/5198) here and added ` <div key={source.name} className="row co-m-form-row">` as a workaround

File: frontend/public/components/utils/details-item.tsx [hunk 6]
```diff
@@ -98,7 +109,9 @@ export const DetailsItem: React.FC<DetailsItemProps> = ({
             <>
               <SplitItem isFilled />
               <SplitItem>
-                <EditButton onClick={onEdit}>{t('public~Edit')}</EditButton>
+                <EditButton label={label} onClick={onEdit}>
+                  {t('public~Edit')}
+                </EditButton>
```
Comment (TheRealJon) [inline frontend/public/components/utils/details-item.tsx:114]:
```suggestion
                <EditButton label={label} onClick={onEdit}>
                  {t('public~Edit')}
                </EditButton>
```

File: frontend/public/components/utils/details-item.tsx [hunk 7]
```diff
@@ -98,7 +109,9 @@ export const DetailsItem: React.FC<DetailsItemProps> = ({
             <>
               <SplitItem isFilled />
               <SplitItem>
-                <EditButton onClick={onEdit}>{t('public~Edit')}</EditButton>
+                <EditButton label={label} onClick={onEdit}>
```
Comment (TheRealJon) [inline frontend/public/components/utils/details-item.tsx:112]: [nit] 'label' is kind of a misleading name for this prop. Maybe `testId` or something?

**Review by TheRealJon — APPROVED** [review]
Just one small nit, but I'm okay to merge this as is.
/lgtm
/approve

**Comment by harishsurf** [comment]
@TheRealJon, `eslint` isn't working for me. I am not sure what is wrong in this [line](https://github.com/openshift/console/blob/3d8051fa179c885337ab4960d465b0fb4f4d5dd8/frontend/public/components/utils/details-item.tsx#L47) as per the [logs](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_console/7191/pull-ci-openshift-console-master-frontend/1347099427952660480/build-log.txt). Could you please check?

**Comment by itsptk** [comment]
> 
> toggle
> 

A later change, but if possible could we change the switches to checkboxes to remain consistent with https://github.com/openshift/openshift-origin-design/issues/392#issuecomment-672835290 ?

**Comment by TheRealJon** [comment]
@harishsurf @spadgett I have the change from switch to checkbox in work.

**Comment by TheRealJon** [comment]
Updated switches to checkboxes:

image

**Comment by TheRealJon** [comment]
@yapei Could you review again since we made some updates?

**Comment by spadgett** [comment]
Should the checkbox label just be the catalog source name instead of enabled/disabled?

**Comment by TheRealJon** [comment]
/hold

For QE approval

**Comment by itsptk** [comment]
> Should the checkbox label just be the catalog source name instead of enabled/disabled?

Agreed, we were thinking maybe something like this for the checkbox version:

image

Thanks all!

**Comment by TheRealJon** [comment]
@itsptk @spadgett @yapei This is ready for one more review. Made the design adjustments and caught some i18n stuff while I was at it.

File: frontend/packages/operator-lifecycle-manager/src/components/operator-hub/operator-hub-details.tsx [hunk 8]
```diff
@@ -6,19 +6,65 @@ import {
...
+import { editDefaultSourcesModal } from '../modals/edit-default-sources-modal';
 
-const OperatorHubDetails: React.FC<OperatorHubDetailsProps> = ({ obj }) => {
+const OperatorHubDetails: React.FC<OperatorHubDetailsProps> = ({ obj: operatorHub }) => {
   const { t } = useTranslation();
+
+  const canEditDefaultSources = useAccessReview({
+    group: OperatorHubModel.apiGroup,
+    resource: OperatorHubModel.plural,
+    verb: 'update',
```
Comment (spadgett) [inline frontend/packages/operator-lifecycle-manager/src/components/operator-hub/operator-hub-details.tsx:23]: Should this be `patch` if the modal uses `k8sPatch`?

Comment (TheRealJon) [reply frontend/packages/operator-lifecycle-manager/src/components/operator-hub/operator-hub-details.tsx:23]: I'm actually not really sure. I'm not intimately familiar with how the access review API works.

Comment (spadgett) [reply frontend/packages/operator-lifecycle-manager/src/components/operator-hub/operator-hub-details.tsx:23]: Yeah, it should be patch if this action spawns the dialog.

**Comment by TheRealJon** [comment]
@spadgett Fixed that access review prop.

**Review by spadgett — APPROVED** [review]
/lgtm

**Comment by spadgett** [comment]
@yanpzhan Can you look again? Thanks!

**Comment by yanpzhan** [comment]
Checked the pr again, now the checkbox is updated, but "Save" text is not marked for translation on the edit model. If it should be tanslated here?
Screenshot from 2021-01-12 13-28-19

**Comment by TheRealJon** [comment]
@yanpzhan Fixed your comment.

**Review by spadgett — APPROVED** [review]
/lgtm

**Comment by spadgett** [comment]
@TheRealJon Needs rebase :/

**Comment by TheRealJon** [comment]
@spadgett Rebased

**Comment by spadgett** [comment]
It looks like this needs some test updates now (and also needs rebase).

**Review by spadgett — APPROVED** [review]
/lgtm
/hold cancel

**Comment by TheRealJon** [comment]
/hold

@spadgett I'm not sure what's happening with this. Looking at the test history, all the tests passed on the last run, but for some reason this retest was still triggered.

**Comment by spadgett** [comment]
@TheRealJon Can you help rebase?

**Review by spadgett — APPROVED** [review]
/lgtm

**Pull request merged**